diff --git a/dwm/config.h b/dwm/config.h
index 585581d..dedd02e 100644
--- a/dwm/config.h
+++ b/dwm/config.h
@@ -172,71 +172,71 @@ static const Key keys[] = {
 	{ MODKEY|ShiftMask,		XK_0,		tag,		{.ui = ~0 } },
 	{ MODKEY,			XK_Tab,		view,		{0} },
 	//{ MODKEY|ShiftMask,		XK_Tab,		spawn,		SHCMD("") },
-	{ MODKEY,			XK_q,		killclient,	{0} },
+	{ MODKEY,			XK_q,		killclient,	{0}, 1 },
 	// { MODKEY|ShiftMask,		XK_q,		quit,		{0} },
-	{ MODKEY|ShiftMask,		XK_q,		spawn,		SHCMD("~/.local/bin/syslock") },
-	{ MODKEY,			XK_w,		spawn,		{.v = (const char*[]){ BROWSER, NULL } } },
-	{ MODKEY|ShiftMask,		XK_w,		spawn,		{.v = (const char*[]){ BROWSER, "--private-window", NULL } } },
-	{ MODKEY|ShiftMask,		XK_p,		spawn,		{.v = (const char*[]){ TERMINAL, "-e", "nmtui", NULL } } },
-	{ MODKEY|ShiftMask,		XK_b,		spawn,		{.v = (const char*[]){ BIT, NULL } } },
-	{ MODKEY,			XK_p,		togglescratch,	{.ui = 3} },
-	{ MODKEY,			XK_Escape,	togglescratch,	{.ui = 2} },
+	{ MODKEY|ShiftMask,		XK_q,		spawn,		SHCMD("~/.local/bin/syslock"), 1 },
+	{ MODKEY,			XK_w,		spawn,		{.v = (const char*[]){ BROWSER, NULL } }, 1 },
+	{ MODKEY|ShiftMask,		XK_w,		spawn,		{.v = (const char*[]){ BROWSER, "--private-window", NULL } }, 1 },
+	{ MODKEY|ShiftMask,		XK_p,		spawn,		{.v = (const char*[]){ TERMINAL, "-e", "nmtui", NULL } }, 1 },
+	{ MODKEY|ShiftMask,		XK_b,		spawn,		{.v = (const char*[]){ BIT, NULL } }, 1 },
+	{ MODKEY,			XK_p,		togglescratch,	{.ui = 3}, 1 },
+	{ MODKEY,			XK_Escape,	togglescratch,	{.ui = 2}, 1 },
 	//{ MODKEY|ShiftMask,		XK_backslash,	spawn,		{.v = (const char*[]){ TERMINAL, "-e", "lf", NULL } } },
-	{ MODKEY,			XK_r,		togglefloating, {0} },
+	{ MODKEY,			XK_r,		togglefloating, {0}, 1 },
 	//{ MODKEY|ShiftMask,		XK_r,		togglefloating,	{0} },
-	{ MODKEY,			XK_t,		setlayout,	{.v = &layouts[0]} }, /* tile */
-	{ MODKEY|ShiftMask,		XK_t,		setlayout,	{.v = &layouts[1]} }, /* bstack */
-	{ MODKEY,			XK_c,		setlayout,	{.v = &layouts[6]} }, /* centeredmaster */
-	{ MODKEY|ShiftMask,		XK_c,		setlayout,	{.v = &layouts[7]} }, /* centeredfloatingmaster */
-	{ MODKEY,			XK_m,		setlayout,	{.v = &layouts[5]} }, /* monocle */
-	{ MODKEY|ShiftMask,		XK_m,		setlayout,	{.v = &layouts[4]} }, /* deck */
-	{ MODKEY,			XK_u,		setlayout,	{.v = &layouts[2]} }, /* spiral */
-	{ MODKEY|ShiftMask,		XK_u,		setlayout,	{.v = &layouts[3]} }, /* dwindle */
+	{ MODKEY,			XK_t,		setlayout,	{.v = &layouts[0]}, 1 }, /* tile */
+	{ MODKEY|ShiftMask,		XK_t,		setlayout,	{.v = &layouts[1]}, 1 }, /* bstack */
+	{ MODKEY,			XK_c,		setlayout,	{.v = &layouts[6]}, 1 }, /* centeredmaster */
+	{ MODKEY|ShiftMask,		XK_c,		setlayout,	{.v = &layouts[7]}, 1 }, /* centeredfloatingmaster */
+	{ MODKEY,			XK_m,		setlayout,	{.v = &layouts[5]}, 1 }, /* monocle */
+	{ MODKEY|ShiftMask,		XK_m,		setlayout,	{.v = &layouts[4]}, 1 }, /* deck */
+	{ MODKEY,			XK_u,		setlayout,	{.v = &layouts[2]}, 1 }, /* spiral */
+	{ MODKEY|ShiftMask,		XK_u,		setlayout,	{.v = &layouts[3]}, 1 }, /* dwindle */
 	//{ MODKEY|ShiftMask,		XK_backslash,	spawn,		SHCMD("") },
-	{ ALTKEY|ShiftMask,		XK_m,		spawn,		SHCMD("~/.local/bin/music.sh") },
-	{ MODKEY,			XK_a,		togglegaps,	{0} },
-	{ MODKEY|ShiftMask,		XK_a,		defaultgaps,	{0} },
-	{ MODKEY,			XK_s,		togglesticky,	{0} },
-	{ MODKEY|ShiftMask,		XK_s,		spawn,		SHCMD("~/.local/bin/displayselect") },
-	{ MODKEY,			XK_d,		spawn,          {.v = (const char*[]){ "dmenu_run", NULL } } },
-	{ MODKEY,			XK_f,		togglefullscr,	{0} },
+	{ ALTKEY|ShiftMask,		XK_m,		spawn,		SHCMD("~/.local/bin/music.sh"), 1 },
+	{ MODKEY,			XK_a,		togglegaps,	{0}, 1 },
+	{ MODKEY|ShiftMask,		XK_a,		defaultgaps,	{0}, 1 },
+	{ MODKEY,			XK_s,		togglesticky,	{0}, 1 },
+	{ MODKEY|ShiftMask,		XK_s,		spawn,		SHCMD("~/.local/bin/displayselect"), 1 },
+	{ MODKEY,			XK_d,		spawn,          {.v = (const char*[]){ "dmenu_run", NULL } }, 1 },
+	{ MODKEY,			XK_f,		togglefullscr,	{0}, 1 },
 	//{ MODKEY|ShiftMask,		XK_f,		setlayout,	{.v = &layouts[8]} },
-	{ MODKEY,			XK_space,	zoom,		{0} },
+	{ MODKEY,			XK_space,	zoom,		{0}, 1 },
 	//{ MODKEY|ShiftMask,		XK_space,	spawn,		{.v = (const char*[]){ "passmenu", NULL } } },
 	{ MODKEY,			XK_g,		shiftview,	{ .i = -1 } },
 	{ MODKEY|ShiftMask,		XK_g,		shifttag,	{ .i = -1 } },
 	{ MODKEY,			XK_h,		setmfact,	{.f = -0.05} },
 	{ MODKEY,			XK_l,		setmfact,      	{.f = +0.05} },
-	{ MODKEY|ShiftMask,		XK_i,		togglescratch,	{.ui = 4} },
-	{ MODKEY,			XK_i,		spawn,		{.v = (const char*[]){ TERMINAL, "-e", "lfub", NULL } } },
+	{ MODKEY|ShiftMask,		XK_i,		togglescratch,	{.ui = 4}, 1 },
+	{ MODKEY,			XK_i,		spawn,		{.v = (const char*[]){ TERMINAL, "-e", "lfub", NULL } }, 1 },
 	{ MODKEY,			XK_semicolon,	shiftview,	{ .i = 1 } },
 	{ MODKEY|ShiftMask,		XK_semicolon,	shifttag,	{ .i = 1 } },
-	{ MODKEY,			XK_apostrophe,	togglescratch,	{.ui = 1} },
-	{ MODKEY|ShiftMask,		XK_apostrophe,	togglesmartgaps,{0} },
-	{ MODKEY,			XK_Return,	spawn,		{.v = termcmd } },
-	{ MODKEY|ShiftMask,		XK_Return,	togglescratch,	{.ui = 0} },
+	{ MODKEY,			XK_apostrophe,	togglescratch,	{.ui = 1}, 1 },
+	{ MODKEY|ShiftMask,		XK_apostrophe,	togglesmartgaps,{0}, 1 },
+	{ MODKEY,			XK_Return,	spawn,		{.v = termcmd }, 1 },
+	{ MODKEY|ShiftMask,		XK_Return,	togglescratch,	{.ui = 0}, 1 },
 	{ MODKEY,			XK_z,		incrgaps,	{.i = +3 } },
 	//{ MODKEY|ShiftMask,		XK_z,		spawn,		SHCMD("") },
 	{ MODKEY,			XK_x,		incrgaps,	{.i = -3 } },
 	//{ MODKEY|ShiftMask,		XK_x,		spawn,		SHCMD("") },
 	//{ MODKEY,			XK_c,		spawn,		SHCMD("") },
-	{ MODKEY,			XK_b,		togglebar,	{0} },
+	{ MODKEY,			XK_b,		togglebar,	{0}, 1 },
 	//{ MODKEY|ShiftMask,		XK_b,		spawn,		SHCMD("") },
-	{ MODKEY,			XK_n,		spawn,		SHCMD("~/.local/bin/notes.sh") },
-	{ MODKEY|ShiftMask,		XK_n,		spawn,		SHCMD(TERMINAL " -e newsboat") },
+	{ MODKEY,			XK_n,		spawn,		SHCMD("~/.local/bin/notes.sh"), 1 },
+	{ MODKEY|ShiftMask,		XK_n,		spawn,		SHCMD(TERMINAL " -e newsboat"), 1 },
 	// { ALTKEY|ShiftMask,		XK_n,		spawn,		SHCMD("~/.local/bin/notes.sh") },
-	{ MODKEY,			XK_e,		spawn,		SHCMD("~/.local/bin/dmenuunicode") },
-	{ MODKEY,			XK_o,		spawn,		SHCMD("xdotool type --delay 0 $(grep -v '^#' ~/.local/share/snippets | dmenu -i -l 10 -p Snippets | cut -d' ' -f1)") },
-	{ MODKEY|ShiftMask,		XK_o,		spawn,		SHCMD("xdotool type --delay 0 $(label=$(grep -v '^#' ~/.local/share/key | awk '{print $2}' | dmenu -i -l 3 -p Key); grep -v '^#' ~/.local/share/key | awk -v pw=\"$label\" '$2 == pw {print $1; exit}')") },
+	{ MODKEY,			XK_e,		spawn,		SHCMD("~/.local/bin/dmenuunicode"), 1 },
+	{ MODKEY,			XK_o,		spawn,		SHCMD("xdotool type --delay 0 $(grep -v '^#' ~/.local/share/snippets | dmenu -i -l 10 -p Snippets | cut -d' ' -f1)"), 1 },
+	{ MODKEY|ShiftMask,		XK_o,		spawn,		SHCMD("xdotool type --delay 0 $(label=$(grep -v '^#' ~/.local/share/key | awk '{print $2}' | dmenu -i -l 3 -p Key); grep -v '^#' ~/.local/share/key | awk -v pw=\"$label\" '$2 == pw {print $1; exit}')"), 1 },
 	/*{ MODKEY|ShiftMask,		XK_s,		spawn,		SHCMD("pkill -USR1 -x sxhkd") },*/
-	{ ALTKEY|ShiftMask,		XK_x,		spawn,		SHCMD("~/.local/bin/ss.sh")},
-	{ ALTKEY,			XK_x,		spawn,		SHCMD("~/.local/bin/tmp-ss.sh")},
-	{ MODKEY|ShiftMask,		XK_x,		spawn,		SHCMD("~/.local/bin/xorg.sh")},
-	{ ALTKEY|ControlMask,		XK_Return,	spawn,		{.v = termcmd } },
-	{ ALTKEY|ControlMask,		XK_p,		spawn,		SHCMD("~/.local/bin/xcompmgr.sh")},
+	{ ALTKEY|ShiftMask,		XK_x,		spawn,		SHCMD("~/.local/bin/ss.sh"), 1 },
+	{ ALTKEY,			XK_x,		spawn,		SHCMD("~/.local/bin/tmp-ss.sh"), 1 },
+	{ MODKEY|ShiftMask,		XK_x,		spawn,		SHCMD("~/.local/bin/xorg.sh"), 1 },
+	{ ALTKEY|ControlMask,		XK_Return,	spawn,		{.v = termcmd }, 1 },
+	{ ALTKEY|ControlMask,		XK_p,		spawn,		SHCMD("~/.local/bin/xcompmgr.sh"), 1 },
 	/*{ ALTKEY|ControlMask,		XK_o,		spawn,		SHCMD("")},*/
-	{ ALTKEY|ShiftMask,		XK_b,		spawn,		SHCMD("~/.local/bin/blue.sh")},
-	{ ALTKEY|ShiftMask,		XK_l,		spawn,		SHCMD("slock")},
+	{ ALTKEY|ShiftMask,		XK_b,		spawn,		SHCMD("~/.local/bin/blue.sh"), 1 },
+	{ ALTKEY|ShiftMask,		XK_l,		spawn,		SHCMD("slock"), 1 },
 	{ ALTKEY,			XK_p,		spawn,		SHCMD("brightnessctl set -e4 5%+") },
 	{ ALTKEY, 			XK_l,		spawn,		SHCMD("brightnessctl set -e4 5%-") },
 	{ 0,				XK_F12,		spawn,		SHCMD("wpctl set-volume @DEFAULT_AUDIO_SINK@ 0%- && wpctl set-volume @DEFAULT_AUDIO_SINK@ 2%+; kill -44 $(pidof dwmblocks)") },
@@ -252,7 +252,7 @@ static const Key keys[] = {
 	{ MODKEY|ShiftMask,		XK_Left,	tagmon,		{.i = -1 } },
 	{ MODKEY,			XK_Right,	focusmon,	{.i = +1 } },
 	{ MODKEY|ShiftMask,		XK_Right,	tagmon,		{.i = +1 } },
-	{ MODKEY,			XK_F5,		xrdb,		{.v = NULL } },
+	{ MODKEY,			XK_F5,		xrdb,		{.v = NULL }, 1 },
 
 	{ 0, XF86XK_MonBrightnessUp,	spawn,		SHCMD("brightnessctl set -e4 5%+") },
 	{ 0, XF86XK_MonBrightnessDown,	spawn,		SHCMD("brightnessctl set -e4 5%-") },
@@ -261,14 +261,13 @@ static const Key keys[] = {
 	{ 0, XF86XK_AudioLowerVolume,	spawn,		SHCMD("wpctl set-volume @DEFAULT_AUDIO_SINK@ 0%+ && wpctl set-volume @DEFAULT_AUDIO_SINK@ 2%-; kill -44 $(pidof dwmblocks)") },
 	{ 0, XF86XK_AudioMicMute,	spawn,		SHCMD("pactl set-source-mute @DEFAULT_SOURCE@ toggle") },
 
-	{ 0, XF86XK_AudioMedia,		spawn,		SHCMD("playerctl play-pause") },
-	{ 0, XF86XK_AudioPlay,		spawn,		SHCMD("playerctl play-pause") },
-	{ 0, XF86XK_AudioPrev,		spawn,		SHCMD("playerctl previous") },
-	{ 0, XF86XK_AudioNext,		spawn,		SHCMD("playerctl next") },
-
-	{ MODKEY,			XK_F11,		spawn,		SHCMD("mpv --untimed --no-cache --no-osc --no-input-default-bindings --profile=low-latency --input-conf=/dev/null --title=webcam $(ls /dev/video[0,2,4,6,8] | tail -n 1)") },
-	{ MODKEY,			XK_BackSpace,	spawn,		SHCMD("~/.local/bin/rec") },
+	{ 0, XF86XK_AudioMedia,		spawn,		SHCMD("playerctl play-pause"), 1 },
+	{ 0, XF86XK_AudioPlay,		spawn,		SHCMD("playerctl play-pause"), 1 },
+	{ 0, XF86XK_AudioPrev,		spawn,		SHCMD("playerctl previous"), 1 },
+	{ 0, XF86XK_AudioNext,		spawn,		SHCMD("playerctl next"), 1 },
 
+	{ MODKEY,			XK_F11,		spawn,		SHCMD("mpv --untimed --no-cache --no-osc --no-input-default-bindings --profile=low-latency --input-conf=/dev/null --title=webcam $(ls /dev/video[0,2,4,6,8] | tail -n 1)"), 1 },
+	{ MODKEY,			XK_BackSpace,	spawn,		SHCMD("~/.local/bin/rec"), 1 },
 };
 
 /* button definitions */
diff --git a/dwm/dwm.c b/dwm/dwm.c
index 09d2b20..b9b7a43 100644
--- a/dwm/dwm.c
+++ b/dwm/dwm.c
@@ -47,6 +47,7 @@
 #endif /* XINERAMA */
 #include <X11/Xft/Xft.h>
 #include <X11/Xlib-xcb.h>
+#include <X11/XKBlib.h>
 #include <xcb/res.h>
 
 #include "drw.h"
@@ -137,6 +138,7 @@ typedef struct {
 	KeySym keysym;
 	void (*func)(const Arg *);
 	const Arg arg;
+	int norepeat;
 } Key;
 
 typedef struct {
@@ -235,6 +237,7 @@ static int gettextprop(Window w, Atom atom, char *text, unsigned int size);
 static void grabbuttons(Client *c, int focused);
 static void grabkeys(void);
 static void keypress(XEvent *e);
+static void keyrelease(XEvent *e);
 static void killclient(const Arg *arg);
 static void manage(Window w, XWindowAttributes *wa);
 static void mappingnotify(XEvent *e);
@@ -347,6 +350,7 @@ static void (*handler[LASTEvent]) (XEvent *) = {
 	[Expose] = expose,
 	[FocusIn] = focusin,
 	[KeyPress] = keypress,
+	[KeyRelease] = keyrelease,
 	[MappingNotify] = mappingnotify,
 	[MapRequest] = maprequest,
 	[PropertyNotify] = propertynotify,
@@ -360,6 +364,7 @@ static Cur *cursor[CurLast];
 static Clr **scheme;
 static Display *dpy;
 static Drw *drw;
+static KeyCode pressedkeycode = 0;
 static Monitor *mons, *selmon;
 static Window root, wmcheckwin;
 
@@ -1324,8 +1329,20 @@ keypress(XEvent *e)
 	for (i = 0; i < LENGTH(keys); i++)
 		if (keysym == keys[i].keysym
 		&& CLEANMASK(keys[i].mod) == CLEANMASK(ev->state)
-		&& keys[i].func)
+		&& keys[i].func) {
+			if (keys[i].norepeat && ev->keycode == pressedkeycode)
+				return;
+			if (keys[i].norepeat)
+				pressedkeycode = ev->keycode;
 			keys[i].func(&(keys[i].arg));
+		}
+}
+
+void
+keyrelease(XEvent *e)
+{
+	if (e->xkey.keycode == pressedkeycode)
+		pressedkeycode = 0;
 }
 
 void
@@ -2009,6 +2026,7 @@ setup(void)
 
 	/* init screen */
 	screen = DefaultScreen(dpy);
+	XkbSetDetectableAutoRepeat(dpy, True, NULL);
 	sw = DisplayWidth(dpy, screen);
 	sh = DisplayHeight(dpy, screen);
 	root = RootWindow(dpy, screen);
