diff --git a/st/config.h b/st/config.h
index e57a1d0..80a2b64 100644
--- a/st/config.h
+++ b/st/config.h
@@ -5,8 +5,8 @@
  *
  * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
  */
-static char *font = "monospace:pixelsize=16:antialias=true:autohint=true";
-static char *font2[] = { "NotoColorEmoji:pixelsize=14:antialias=true:autohint=true" };
+static char *font = "monospace:pixelsize=12:antialias=true:autohint=true";
+static char *font2[] = { "NotoColorEmoji:pixelsize=10:antialias=true:autohint=true" };
 static int borderpx = 0;
 
 /*
@@ -109,34 +109,34 @@ char *termname = "st-256color";
 unsigned int tabspaces = 8;
 
 /* bg opacity */
-float alpha = 1.0;
+float alpha = 0.8;
 float alphaOffset = 0.0;
 float alphaUnfocus;
 
 /* Terminal colors (16 first used in escape sequence) */
 static const char *colorname[] = {
-	"#ffffff", /* hard contrast: #ffffff / soft contrast: #f8f8f8 */
-	"#a60000",
-	"#006800",
-	"#6f5500",
-	"#0031a9",
-	"#721045",
-	"#005e8b",
-	"#bfc0c0",
-	"#505050",
-	"#972500",
-	"#316500",
-	"#884900",
-	"#354fcf",
-	"#531ab6",
-	"#005f5f",
-	"#000000",
+	"#282828", /* hard contrast: #1d2021 / soft contrast: #32302f */
+	"#cc241d",
+	"#98971a",
+	"#d79921",
+	"#458588",
+	"#b16286",
+	"#689d6a",
+	"#a89984",
+	"#928374",
+	"#fb4934",
+	"#b8bb26",
+	"#fabd2f",
+	"#83a598",
+	"#d3869b",
+	"#8ec07c",
+	"#ebdbb2",
 	[255] = 0,
 	/* more colors can be added after 255 to use with DefaultXX */
-	"#000000", /* 256 -> cursor */
-	"#ffffff", /* 257 -> rev cursor */
-	"#ffffff", /* 258 -> bg */
-	"#000000", /* 259 -> fg */
+	"#add8e6", /* 256 -> cursor */
+	"#555555", /* 257 -> rev cursor*/
+	"#282828", /* 258 -> bg */
+	"#ebdbb2", /* 259 -> fg */
 };
 
 
